### ISSUE TITLE
Type wildcard

### ISSUE THREAD

**Issue by yannham** [tweag/nickel#495]
**Is your feature request related to a problem? Please describe.**
In Nickel, an annotation is required to trigger static typechecking. But

1. Sometimes, we may have no clue about the type, and might want the typechecker to infer it for us. Say a program is failing and we don't really understand why, but there's a problem with the value of `foo`. We might be tempted to call the static typechecker to the rescue, but we may not know what should be the type to but there:
    ```
    let foo : ? = ... in
    ```

2. Sometimes, we don't want to repeat ourselves. A common example is when using a record of functions as a library, when we might already have annotated the individual elements:
    ```
    {
      foo : Foo -> Bar = ...,
      bar : Bar -> Foo = ....,
    }
    ````
    By virtue of Nickel's gradual typing semantics, the top level record is not annotated, and is thus affected to type `Dyn`. All the typing information is lost, and one can't use `lib.foo` in a statically typed context correctly. Currently, one way is to repeat all the type signatures at the record level:
    ```
    {
      foo : Foo -> Bar = ...,
      bar : Bar -> Foo = ...,
    } : {
      foo : Foo -> Bar,
      bar : Bar -> Foo,
    }
    ```
    But this is tedious. We can also get rid of the original signature on each function, and only keep the top level one, but it makes reading the code harder (we don't see at each function definition what is the signature).

**Describe the solution you'd like**
As decided during the [standardization meeting](https://github.com/tweag/nickel/blob/master/notes/standardization-meeting-07.12.21.md), both cases could be solved by type wildcards. A wildcard is denoted by `_`, and means exactly "please infer this type for me". It could be used as any type, inside other types: `let bar : _ = in ...`, `let foo : _ -> _ -> Num = ...` etc.

The semantics would be that the typechecker replaces each wildcard by a fresh unification variable. Once typechecking is over, the wildcard is overwritten by a concrete type, so that runtime contract checks can proceed normally. Any unresolved variable is replaced by `Dyn`. For example:

```
// Get type `Dyn -> Dyn`
let f : _ = fun x => x in ...
```

**Comment by yannham** [comment]
Notes: Haskell has a slightly different type hole feature, more related to interactive programming (the typechecker reports the type it has found for the hole). If we are to have this too, we should probably use a different syntax, like `?`.

**Comment by aspiwack** [comment]
I suggest you call this here feature a “type wildcard”.

**Comment by yannham** [comment]
Closed by #664. We may revisit later if we want the inferred types to be accessible from other blocks, but that's a backward compatible improvement (That is, we implemented something that only solves 1. but not 2., to keep typechecking modular)
